# HIVmicroglia
Script related to the microglia HIV paper


## Explanations  

### IS locations, profiles of IS genes (comparison with ChIP-seq and ATAC-seq) 
- `Figure1_SuppFig1.Rmd` : IS locations assessment, expression comparisons, and GO enrichment (All panels from Figure 1 and panels D-F in supplementary figure 1)
- `Figure2_Supp2A-C.Rmd` : Metagene plots by expression levels (Panels A-C from Figures 2 and Supplementary 2)
- `Figure2D.Rmd` : Profile plots for IS and RMC (Panel D from Figure 2) 
- `SuppFigure2D.Rmd` : Profile plots for IS and RMC by genic/intergenic (Panel D from Supplementary Figure 2)
- `Figure2E.Rmd` : Comparison with the ChromHMM model  

### NMF-derived analysis  
- `Figure3AB.Rmd` : NMF-analysis derived plots for the H and W matrix


### ATAC-seq processing

- `2021-11-02_atacseq.Rmd` : Quality control / annotation of ATAC-seq peaks
- `2021-11-04_diffATAC.Rmd` : Differential ATAC-seq analysis between 3 conditions

### TF analysis
- `Figure4.Rmd` : CTCF dynamics and footprinting (Fig 4D and Supp Fig 5C), along with heatmap on feature importance compared with TF binding by condition (panel 4E) 

